mars/services/scheduling/supervisor/manager.py

```diff
@@ -56,35 +68,150 @@ def to_summary(self, **kwargs) -> SubtaskScheduleSummary:
 class SubtaskManagerActor(mo.Actor):
     _subtask_infos: Dict[str, SubtaskScheduleInfo]  # subtask id -> schedule info
     _subtask_summaries: Dict[str, SubtaskScheduleSummary]  # subtask id -> summary
+    _grouped_unfinished_subtasks: Dict[str, Dict[str, SubtaskScheduleInfo]]  # key is subtask logic id
+    _grouped_finished_subtasks: Dict[str, Dict[str, SubtaskScheduleInfo]]  # key is subtask logic id
+    _speculation_execution_task: Optional[asyncio.Task]
 
     @classmethod
     def gen_uid(cls, session_id: str):
-        return f"{session_id}_subtask_manager"
+        return f'{session_id}_subtask_manager'
 
-    def __init__(
-        self,
-        session_id: str,
-        subtask_max_reschedules: int = DEFAULT_SUBTASK_MAX_RESCHEDULES,
-    ):
+    def __init__(self,
+                 session_id: str,
+                 subtask_max_reschedules: int = DEFAULT_SUBTASK_MAX_RESCHEDULES,
+                 subtask_cancel_timeout: int = 5,
+                 speculation_config: Dict[str, object] = None):
         self._session_id = session_id
         self._subtask_infos = dict()
         self._subtask_summaries = dict()
+        self._grouped_unfinished_subtasks = defaultdict(dict)
```

Comment:
Putting all speculation strategy code in `SubtaskManagerActor` directly seems too heavy for me any this may not be easy to write unit tests for these strategies. Is it possible to create some abstractions?